src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -103,10 +97,11 @@ public virtual void Dispose()
                 current = default(TSource);
                 state = -1;
             }
-
+            
             public IEnumerator<TSource> GetEnumerator()
             {
-                Iterator<TSource> enumerator = state == 0 && _threadId == Environment.CurrentManagedThreadId ? this : Clone();
+                if (Interlocked.CompareExchange(ref state, 1, 0) == 0) return this;
```

Comment:
I was just going by rule 1 in the [Coding Styles](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md) document.